crates/swc_ecma_lints/src/rules/no_use_before_define.rs

```diff
@@ -0,0 +1,233 @@
+            }
+            LintRuleReaction::Warning => {
+                handler.struct_span_warn(span, &message).emit();
+            }
+            _ => {}
+        });
+    }
+
+    fn check_ident(&self, es6_var_check: bool, ident: &Ident) {
+        let ident_id = ident.to_id();
+
+        if self.has_access_before_define(&ident_id) {
+            let sym = &*ident.sym;
+            let mut spans = self
+                .get_ident_spans_in_current_scope(&ident_id)
+                .iter()
+                .copied()
+                .collect::<Vec<Span>>();
+
+            spans.sort();
+
+            spans.into_iter().for_each(|span| {
+                self.emit_report(es6_var_check, span, sym);
+            });
+        }
+    }
+
+    fn check_pat(&self, es6_var_check: bool, pat: &Pat) {
+        match pat {
+            Pat::Ident(BindingIdent { id, .. }) => {
+                self.check_ident(es6_var_check, id);
+            }
+            Pat::Array(ArrayPat { elems, .. }) => {
+                elems.iter().for_each(|elem| {
+                    if let Some(elem) = elem {
+                        self.check_pat(es6_var_check, elem);
+                    }
+                });
+            }
+            Pat::Object(ObjectPat { props, .. }) => {
+                props.iter().for_each(|prop| match prop {
+                    ObjectPatProp::Assign(AssignPatProp { key, .. }) => {
+                        self.check_ident(es6_var_check, key);
+                    }
+                    ObjectPatProp::KeyValue(KeyValuePatProp { value, .. }) => {
+                        self.check_pat(es6_var_check, value.as_ref());
+                    }
+                    ObjectPatProp::Rest(RestPat { arg, .. }) => {
+                        self.check_pat(es6_var_check, arg.as_ref());
+                    }
+                });
+            }
+            Pat::Rest(RestPat { arg, .. }) => {
+                self.check_pat(es6_var_check, arg.as_ref());
+            }
+            Pat::Assign(AssignPat { left, .. }) => {
+                self.check_pat(es6_var_check, left.as_ref());
+            }
+            Pat::Invalid(_) => {}
+            Pat::Expr(_) => {}
+        }
+    }
+}
+
+impl Visit for NoUseBeforeDefine {
+    noop_visit_type!();
+
+    fn visit_ident(&mut self, ident: &Ident) {
+        self.insert_ident_to_current_scope(ident.to_id(), ident.span);
+    }
+
+    fn visit_block_stmt(&mut self, block: &BlockStmt) {
+        self.scoped_indents.insert(block.span, Default::default());
+        self.scope.push(block.span);
+
+        block.visit_children_with(self);
+
+        self.scoped_indents.remove(&block.span);
+        self.scope.pop();
+    }
+
+    fn visit_var_decl(&mut self, var_decl: &VarDecl) {
+        let es6_var_check = !matches!(&var_decl.kind, VarDeclKind::Var);
+
+        var_decl.decls.iter().for_each(|declarator| {
+            if let Some(init) = &declarator.init {
```

Comment:
I think`declarator.init.visit_with(self)` is better, without `if let Some()`.

Reply:
Description: Behavior is identical, as

 - `visit_opt_expr` is not defined. `if let Some` part is done by default impl of `visit_opt_expr`.
 - `visit_expr` is not defined. `visit_children_with(self)` is done by default impl of `visit_expr`.